### PR TITLE
Purification

### PR DESCRIPTION
**Pull request by vakaras**
A replacement of #369. I will close that one once this is merged.

* [x] Add an action that runs some tests with `ENABLE_PURIFICATION_OPTIMIZATION = true`.

### CODE REVIEW

File: prusti-common/src/config.rs [hunk 1]
```diff
@@ -379,6 +380,10 @@ pub fn optimizations() -> Optimizations {
     return opt;
 }
 
+pub fn enable_purification_optimization() -> bool {
```
Comment (Aurel300) [inline prusti-common/src/config.rs]: Should have a short comment explanation like the other functions.

File: prusti-tests/tests/verify/pass/equality/quantifiers.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+use prusti_contracts::*;
+
+#[derive(PartialEq, Eq)]
+struct A { x: usize }
+
+#[requires(forall(|other: A| _foo == other))]
+fn test(_foo: A) {}
+
+#[requires(forall(|other: u32| _foo == other))]
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/equality/quantifiers.rs]: For `verify/pass`, this is quite a strange test case. As a user I would expect this to fail.

File: prusti-viper/src/encoder/builtin_encoder.rs [hunk 3]
```diff
@@ -88,4 +94,65 @@ impl BuiltinEncoder {
...
+            functions.push(f.clone());
+
+            let forall_arg = vir::LocalVar {
+                name: "self".to_owned(),
+                typ: t.clone(),
+            };
+            let function_app =
+                vir::Expr::domain_func_app(f.clone(), vec![vir::Expr::local(forall_arg.clone())]);
+            let body = vir::Expr::and(function_app, true.into());
+            let e = vir::Expr::forall(vec![forall_arg], vec![], body); //TODO triggers
```
Comment (Aurel300) [inline prusti-viper/src/encoder/builtin_encoder.rs]: What triggers should this have?

File: prusti-viper/src/encoder/builtin_encoder.rs [hunk 4]
```diff
@@ -88,4 +94,65 @@ impl BuiltinEncoder {
...
+            let ax = vir::DomainAxiom {
+                name: format!("{}$axiom", f.get_identifier()),
+                expr: e,
+                domain_name: domain_name.to_string(),
+            };
+            axioms.push(ax); //TODO
```
Comment (Aurel300) [inline prusti-viper/src/encoder/builtin_encoder.rs]: What is there to do?

File: prusti-viper/src/encoder/encoder.rs [hunk 5]
```diff
@@ -252,10 +263,144 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
...
+            let builtin_encoder =  BuiltinEncoder::new();
+            domains.push(builtin_encoder.encode_builtin_domain(BuiltinDomainKind::Nat));
+            domains.push(builtin_encoder.encode_builtin_domain(BuiltinDomainKind::Primitive));
+        }
+
         domains.sort_by_key(|d| d.get_identifier());
         domains
     }
 
+    pub fn encode_axiomatized_pure_function(&self, f: &vir::Function) {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/encoder.rs]: This is a huge function that could use some simplifying. Does it actually belong to `Encoder`?

File: prusti-viper/src/encoder/encoder.rs [hunk 6]
```diff
@@ -252,10 +263,144 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
...
+                    None
+                } else {
+                    Some(vir::ExprFolder::fold(&mut purifier, p))
+                }
+            })
+            .conjoin();
+
+        let function_body = vir::ExprFolder::fold(&mut purifier, f.body.clone().unwrap());
+
+        let function_identiry = vir::Expr::eq_cmp(function_call_with_succ.clone(), function_body);
```
Comment (Aurel300) [inline prusti-viper/src/encoder/encoder.rs]:
```suggestion
        let function_identity = vir::Expr::eq_cmp(function_call_with_succ.clone(), function_body);
```

Also (incl. this one) many cases of unnecessary variables that are just used once.

File: prusti-viper/src/encoder/encoder.rs [hunk 7]
```diff
@@ -1141,6 +1296,17 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
...
             };
 
+            if config::enable_purification_optimization() {
+                // Ensure that snapshots of all types used in the function are
+                // already encoded.
+                let body = procedure.get_mir();
+                for local_decl in &body.local_decls {
+                    let ty = local_decl.ty;
+                    self.encode_snapshot(ty).with_span(procedure.get_span()).run_if_err(cleanup)?;
+                }
```
Comment (Aurel300) [inline prusti-viper/src/encoder/encoder.rs]: Shouldn't this be triggered "on demand" in `encode_axiomatized_pure_function`?

File: prusti-viper/src/encoder/snapshot/fixer.rs [hunk 8]
```diff
@@ -0,0 +1,37 @@
+use std::collections::HashMap;
+use prusti_common::vir::{self, ExprFolder};
+use crate::encoder::snapshot_encoder::Snapshot;
+
+pub struct Fixer<'a> {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/snapshot/fixer.rs:6]: Same idea is in `SnapshotSpecPatcher`. Why "fixer"? Can the two not be merged?

File: prusti-viper/src/encoder/snapshot/mod.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
+use std::collections::HashMap;
+
+use log::{info, warn};
+use prusti_common::vir;
+
+pub use self::purifier::ExprPurifier;
+
+use super::{errors::EncodingResult, snapshot_encoder::Snapshot};
+use vir::Type;
```
Comment (Aurel300) [inline prusti-viper/src/encoder/snapshot/mod.rs]: Why this type specifically? The file uses `vir::Type` in many (most?) places anyway.

**Comment by Aurel300** [comment]
Left some comments. Ignoring most of the snapshot encoder because that is mostly replaced by [`cleanup-snapshot`](https://github.com/viperproject/prusti-dev/compare/master...Aurel300:cleanup-snapshot).

File: prusti-common/src/config.rs (hunk 1 above)
Comment (vakaras) [reply prusti-common/src/config.rs]: Added.

File: prusti-tests/tests/verify/pass/equality/quantifiers.rs (hunk 2 above)
Comment (vakaras) [reply prusti-tests/tests/verify/pass/equality/quantifiers.rs]: It seems that Github got confused. This test is not part of the PR.

File: prusti-viper/src/encoder/builtin_encoder.rs (hunk 3 above)
Comment (vakaras) [reply prusti-viper/src/encoder/builtin_encoder.rs]: Added triggers.

File: prusti-viper/src/encoder/builtin_encoder.rs (hunk 4 above)
Comment (vakaras) [reply prusti-viper/src/encoder/builtin_encoder.rs]: Should be fixed.

File: prusti-viper/src/encoder/encoder.rs (hunk 5 above)
Comment (vakaras) [reply prusti-viper/src/encoder/encoder.rs]: Move the function to a separate file and simplified a bit.

File: prusti-viper/src/encoder/snapshot/fixer.rs (hunk 8 above)
Comment (vakaras) [reply prusti-viper/src/encoder/snapshot/fixer.rs:6]: Added a comment that explains the problem.

File: prusti-viper/src/encoder/snapshot/mod.rs (hunk 9 above)
Comment (vakaras) [reply prusti-viper/src/encoder/snapshot/mod.rs]: Removed the type.

**Comment by vakaras** [comment]
@Aurel300 I am planning to merge this once the CI is green.

**Comment by vakaras** [comment]
@tillarnold The Fibonacci example was failing to verify for me due to a bug, which I fixed. Could you please rerun your benchmarks on the master branch to check whether the fix has any influence on the performance? Also, which tests are supposed to work with verification enabled? I would like to add them to CI so that we do not get regressions by accident.

**Comment by tillarnold** [comment]
@vakaras yes your fix appears to have significantly improved the performance of Fibonacci. It went from from 14 down to 5 seconds. Which is still slower than the non purified version with 3 seconds, but it is a significant improvement.  The other benchmarks appear to be similar as before. 

The files in  `benchmarked-files.csv` are supposed to work. That would be:

<details> 

```
prusti-tests/tests/verify/pass/rosetta/Knuth_shuffle.rs
prusti-tests/tests/verify/pass/demos/account.rs
prusti-tests/tests/verify/pass/rosetta/Knights_tour.rs
prusti-tests/tests/verify/pass/quick/fold-unfold.rs
prusti-tests/tests/verify/pass/quick/moves.rs
prusti-tests/tests/verify/pass/quick/mut-borrows.rs
prusti-tests/tests/verify/pass/quick/shared-borrows.rs
prusti-tests/tests/verify/pass/quick/trait-contracts-refinement.rs
prusti-tests/tests/verify/pass/quick/routes.rs
prusti-tests/tests/verify/pass/quick/fibonacci.rs
prusti-tests/tests/verify/pass/pure-fn/len-lookup.rs
prusti-tests/tests/verify/pass/pure-fn/quantifiers.rs
prusti-tests/tests/verify/pass/pure-fn/recursive-pure-fn.rs
prusti-tests/tests/verify/pass/pure-fn/ref-mut-arg.rs
prusti-tests/tests/verify/pass/rosetta/Ackermann_function.rs
prusti-tests/tests/verify/pass/rosetta/Heapsort.rs
```

</details> 

It however appears that routes.rs no longer works.

**Comment by vakaras** [comment]
> It however appears that routes.rs no longer works.

It is strange that it worked before because to my understanding it requires UnfoldWitness'es.

**Pull request merged**